lib/puppet/util/execution.rb

```diff
@@ -216,28 +228,32 @@ def self.execute(command, options = NoOptionsSpecified)
       elsif Puppet.features.microsoft_windows?
         process_info = execute_windows(*exec_args)
         begin
+          [stdin, stdout, stderr].each {|io| io.close rescue nil}
+          unless options[:squelch]
+            while !reader.eof?
+              output << reader.read
+            end
+            read_succeeded = true
+          end
           exit_status = Puppet::Util::Windows::Process.wait_process(process_info.process_handle)
         ensure
           FFI::WIN32.CloseHandle(process_info.process_handle)
           FFI::WIN32.CloseHandle(process_info.thread_handle)
         end
       end
 
-      [stdin, stdout, stderr].each {|io| io.close rescue nil}
-
-      # read output in if required
-      unless options[:squelch]
-        output = wait_for_output(stdout)
-        Puppet.warning "Could not get output" unless output
+      unless options[:squelch] || read_succeeded
```

Comment:
Is `read_succeeded ` necessary? If `options[:squelch]` is `false`, then we always call (for posix and windows):

```ruby
while !reader.eof?
  output << reader.read
end
read_succeeded = true
```

which always sets `read_succeeded ` to `true` or will raise. But if it's the latter, then we never reach this line.